File: .github/workflows/lib/approval.js
class ApprovalManager {
  constructor(github, org, repo, issueNumber) {
    this.github = github;
    this.org = org;
    this.repo = repo;
    this.issueNumber = issueNumber;
    this.coreTeamMembers = [];
    this.maintainerTeamMembers = [];
    this.comments = [];
    this.coreApprovals = new Set();
    this.maintainerApprovals = new Set();
    this.coreRejections = new Set();
    this.maintainerRejections = new Set();
    this.awaitingCore = [];
    this.awaitingMaintainers = [];
  }

  // Helper for list formatting
  formatUserList(users) {
    return users.length ? users.map((u) => `[@${u}](https://github.com/${u})`).join(", ") : "-";
  }

  // Determine the status of a pipeline proposal from durable issue state and
  // votes, rather than from the event that happened to trigger the workflow.
  determinePipelineStatus(issue) {
    const labels = (issue.labels || []).map((label) => (typeof label === "string" ? label : label.name));
    const approvalThresholdReached =
      this.coreApprovals.size >= 2 || (this.coreApprovals.size >= 1 && this.maintainerApprovals.size >= 1);
    const rejectionThresholdReached =
      (this.coreRejections.size >= 2 || (this.coreRejections.size >= 1 && this.maintainerRejections.size >= 1)) &&
      this.coreApprovals.size === 0 &&
      this.maintainerApprovals.size === 0;
    const closedAsRejected =
      issue.state === "closed" &&
      issue.state_reason === "not_planned" &&
      (this.coreRejections.size > 0 || this.maintainerRejections.size > 0);

    if (labels.includes("timed-out")) {
      return "⏰ Timed Out";
    }
    if (closedAsRejected || rejectionThresholdReached) {
      return "❌ Rejected";
    }
    if (approvalThresholdReached) {
      return "✅ Approved";
    }
    return "🕐 Pending";
  }

  // Helper to fetch team members
  async getTeamMembers(teamSlug) {
    try {
      const res = await this.github.request("GET /orgs/{org}/teams/{team_slug}/members", {
        org: this.org,
        team_slug: teamSlug,
        per_page: 100,
      });
      console.log(`Fetched ${res.data.length} ${teamSlug} team members.`);
      return res.data.map((m) => m.login);
    } catch (err) {
      console.error(`Failed to fetch ${teamSlug} team members:`, err);
      throw err;
    }
  }

  // Initialize the manager with team members and comments
  async initialize() {
    // Fetch team members
    this.coreTeamMembers = await this.getTeamMembers("core");
    this.maintainerTeamMembers = await this.getTeamMembers("maintainers");
    console.log("Core team members:", this.coreTeamMembers);
    console.log("Maintainer team members:", this.maintainerTeamMembers);

    // Fetch comments
    this.comments = await this.github.paginate(this.github.rest.issues.listComments, {
      owner: this.org,
      repo: this.repo,
      issue_number: this.issueNumber,
      per_page: 100,
    });

    // Process comments
    this.processComments();
    return this;
  }

  // Helper to update issue status and labels
  async updateIssueStatus(status) {
    // Get current issue to preserve existing labels
    const issue = await this.github.rest.issues.get({
      owner: this.org,
      repo: this.repo,
      issue_number: this.issueNumber,
    });

    // Filter out existing status labels
    const statusLabels = ["accepted", "turned-down", "timed-out", "proposed"];
    const existingLabels = issue.data.labels
      .map((label) => (typeof label === "string" ? label : label.name))
      .filter((label) => !statusLabels.includes(label));

    // Determine new status label
    let newStatusLabel;
    switch (status) {
      case "✅ Approved":
        newStatusLabel = "accepted";
        break;
      case "❌ Rejected":
        newStatusLabel = "turned-down";
        break;
      case "⏰ Timed Out":
        newStatusLabel = "timed-out";
        break;
      default:
        newStatusLabel = "proposed";
    }

    // Combine existing non-status labels with new status label
    const updatedLabels = [...existingLabels, newStatusLabel];

    // Avoid triggering another workflow run when the labels are already right.
    const currentLabels = issue.data.labels.map((label) => (typeof label === "string" ? label : label.name));
    if (
      currentLabels.length === updatedLabels.length &&
      currentLabels.every((label) => updatedLabels.includes(label))
    ) {
      console.log("Issue labels already up to date - no update required.");
      return;
    }

    // Update labels
    await this.github.rest.issues.update({
      owner: this.org,
      repo: this.repo,
      issue_number: this.issueNumber,
      labels: updatedLabels,
    });
  }

  // Helper to find and update status comment
  async updateStatusComment(statusBody) {
    const statusComments = this.comments.filter((c) => c.body && c.body.startsWith("## Approval status:"));
    const statusComment = statusComments[0];

    // Opening an issue also applies its template labels. Older workflow runs
    // could race and create more than one status comment; clean those up when
    // the issue is next processed.
    for (const duplicate of statusComments.slice(1)) {
      console.log(`Deleting duplicate status comment ${duplicate.id}.`);
      await this.github.rest.issues.deleteComment({
        owner: this.org,
        repo: this.repo,
        comment_id: duplicate.id,
      });
    }

    if (statusComment) {
      if (statusComment.body.trim() === statusBody.trim()) {
        console.log("Status comment already up to date - no update required.");
      } else {
        console.log("Updating existing status comment.");
        await this.github.rest.issues.updateComment({
          owner: this.org,
          repo: this.repo,
          comment_id: statusComment.id,
          body: statusBody,
        });
      }
    } else {
      // Fallback: create a new status comment if missing
      await this.github.rest.issues.createComment({
        owner: this.org,
        repo: this.repo,
        issue_number: this.issueNumber,
        body: statusBody,
      });
    }
  }

  // Helper to process comments and collect votes
  processComments() {
    // Reset all approval sets
    this.coreApprovals = new Set();
    this.maintainerApprovals = new Set();
    this.coreRejections = new Set();
    this.maintainerRejections = new Set();

    for (const comment of this.comments) {
      const commenter = comment.user.login;
      const isCoreMember = this.coreTeamMembers.includes(commenter);
      const isMaintainer = this.maintainerTeamMembers.includes(commenter);

      if (!isCoreMember && !isMaintainer) continue; // Only team members count

      // Skip comments with no body
      if (!comment.body) continue;

      // Count approvals / rejections based on line starting with /approve or /reject
      const lines = comment.body.split(/\r\n|\r|\n/);
      for (const rawLine of lines) {
        const line = rawLine.trim();
        if (/^\/approve\b/i.test(line)) {
          if (isCoreMember) {
            this.coreApprovals.add(commenter);
            this.coreRejections.delete(commenter); // Remove any previous rejection
          } else if (isMaintainer) {
            this.maintainerApprovals.add(commenter);
            this.maintainerRejections.delete(commenter); // Remove any previous rejection
          }
        } else if (/^\/reject\b/i.test(line)) {
          if (isCoreMember) {
            this.coreRejections.add(commenter);
            this.coreApprovals.delete(commenter); // Remove any previous approval
          } else if (isMaintainer) {
            this.maintainerRejections.add(commenter);
            this.maintainerApprovals.delete(commenter); // Remove any previous approval
          }
        }
      }
    }

    // Update awaiting lists
    this.awaitingCore = this.coreTeamMembers.filter((u) => !this.coreApprovals.has(u) && !this.coreRejections.has(u));
    this.awaitingMaintainers = this.maintainerTeamMembers.filter(
      (u) => !this.maintainerApprovals.has(u) && !this.maintainerRejections.has(u),
    );
  }
}

module.exports = ApprovalManager;

File: .github/workflows/lib/approval.py
import re
import os
from typing import Set, List, Optional
from github import Github
from github.Repository import Repository
from github.Issue import Issue


class ApprovalManager:
    def __init__(self, github_token: str, org: str, repo: str, issue_number: int):
        self.github = Github(github_token)
        self.org = org
        self.repo = repo
        self.issue_number = issue_number
        self.repository: Repository = self.github.get_repo(f"{org}/{repo}")
        self.issue: Issue = self.repository.get_issue(issue_number)

        self.core_team_members: List[str] = []
        self.maintainer_team_members: List[str] = []
        self.comments = []
        self.core_approvals: Set[str] = set()
        self.maintainer_approvals: Set[str] = set()
        self.core_rejections: Set[str] = set()
        self.maintainer_rejections: Set[str] = set()
        self.awaiting_core: List[str] = []
        self.awaiting_maintainers: List[str] = []

    def format_user_list(self, users: List[str]) -> str:
        """Helper for list formatting"""
        if not users:
            return "-"
        return ", ".join([f"[@{user}](https://github.com/{user})" for user in users])

    def get_team_members(self, team_slug: str) -> List[str]:
        """Helper to fetch team members"""
        try:
            org = self.github.get_organization(self.org)
            team = org.get_team_by_slug(team_slug)
            members = list(team.get_members())
            print(f"Fetched {len(members)} {team_slug} team members.")
            return [member.login for member in members]
        except Exception as err:
            print(f"Failed to fetch {team_slug} team members: {err}")
            raise err

    def initialize(self):
        """Initialize the manager with team members and comments"""
        # Fetch team members
        self.core_team_members = self.get_team_members("core")
        self.maintainer_team_members = self.get_team_members("maintainers")
        print("Core team members:", self.core_team_members)
        print("Maintainer team members:", self.maintainer_team_members)

        # Fetch comments
        self.comments = list(self.issue.get_comments())

        # Process comments
        self.process_comments()
        return self

    def update_issue_status(self, status: str):
        """Helper to update issue status and labels"""
        # Get current labels
        current_labels = [label.name for label in self.issue.labels]

        # Filter out existing status labels
        status_labels = ["accepted", "turned-down", "timed-out", "proposed"]
        existing_labels = [label for label in current_labels if label not in status_labels]

        # Determine new status label
        status_map = {
            "✅ Approved": "accepted",
            "❌ Rejected": "turned-down",
            "⏰ Timed Out": "timed-out"
        }
        new_status_label = status_map.get(status, "proposed")

        # Combine existing non-status labels with new status label
        updated_labels = existing_labels + [new_status_label]

        # Update labels
        self.issue.set_labels(*updated_labels)

    def update_status_comment(self, status_body: str):
        """Helper to find and update status comment"""
        status_comment = None
        for comment in self.comments:
            if comment.body.startswith("## Approval status:"):
                status_comment = comment
                break

        if status_comment:
            if status_comment.body.strip() == status_body.strip():
                print("Status comment already up to date - no update required.")
            else:
                print("Updating existing status comment.")
                status_comment.edit(status_body)
        else:
            # Fallback: create a new status comment if missing
            self.issue.create_comment(status_body)

    def process_comments(self):
        """Helper to process comments and collect votes"""
        # Reset all approval sets
        self.core_approvals = set()
        self.maintainer_approvals = set()
        self.core_rejections = set()
        self.maintainer_rejections = set()

        for comment in self.comments:
            commenter = comment.user.login
            is_core_member = commenter in self.core_team_members
            is_maintainer = commenter in self.maintainer_team_members

            if not is_core_member and not is_maintainer:
                continue  # Only team members count

            # Skip comments with no body
            if not comment.body:
                continue

            # Count approvals / rejections based on line starting with /approve or /reject
            lines = comment.body.split('\n')
            for line in lines:
                line = line.strip()
                if re.match(r'^/approve\b', line, re.IGNORECASE):
                    if is_core_member:
                        self.core_approvals.add(commenter)
                        self.core_rejections.discard(commenter)  # Remove any previous rejection
                    elif is_maintainer:
                        self.maintainer_approvals.add(commenter)
                        self.maintainer_rejections.discard(commenter)  # Remove any previous rejection
                elif re.match(r'^/reject\b', line, re.IGNORECASE):
                    if is_core_member:
                        self.core_rejections.add(commenter)
                        self.core_approvals.discard(commenter)  # Remove any previous approval
                    elif is_maintainer:
                        self.maintainer_rejections.add(commenter)
                        self.maintainer_approvals.discard(commenter)  # Remove any previous approval

        # Update awaiting lists
        self.awaiting_core = [
            user for user in self.core_team_members
            if user not in self.core_approvals and user not in self.core_rejections
        ]
        self.awaiting_maintainers = [
            user for user in self.maintainer_team_members
            if user not in self.maintainer_approvals and user not in self.maintainer_rejections
        ]
